cartographer/cloud/internal/local_trajectory_uploader.cc

```diff
@@ -143,9 +202,17 @@ void LocalTrajectoryUploader::ProcessSendQueue() {
 
       if (batch_request.sensor_data_size() == batch_size_) {
         async_grpc::Client<handlers::AddSensorDataBatchSignature> client(
-            client_channel_, async_grpc::CreateUnlimitedConstantDelayStrategy(
-                                 common::FromSeconds(1)));
-        CHECK(client.Write(batch_request));
+            client_channel_,
+            async_grpc::CreateUnlimitedConstantDelayStrategy(
+                common::FromSeconds(1), kUnrecoverableStatusCodes));
+        LOG(INFO) << "Uploading";
+        if (!client.Write(batch_request)) {
+          // Unrecoverable error occurred.
+          batch_request.clear_sensor_data();
+          TryRecovery();
+        } else {
```

Comment:
"if (positive) ... continue; ..." may be more readable than "if (negative) ... else ..."

Reply:
done